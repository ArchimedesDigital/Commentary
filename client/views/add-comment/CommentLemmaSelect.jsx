import RaisedButton from 'material-ui/RaisedButton';
import FontIcon from 'material-ui/FontIcon';
import TextField from 'material-ui/TextField';
import baseTheme from 'material-ui/styles/baseThemes/lightBaseTheme';
import getMuiTheme from 'material-ui/styles/getMuiTheme';

CommentLemmaSelect = React.createClass({

    propTypes: {
      workSlug: React.PropTypes.string.isRequired,
      subworkN: React.PropTypes.number.isRequired,
      selectedLineFrom: React.PropTypes.number.isRequired,
      selectedLineTo: React.PropTypes.number.isRequired,
    },

    childContextTypes: {
      muiTheme: React.PropTypes.object.isRequired,
    },

    getChildContext() {
      return {muiTheme: getMuiTheme(baseTheme)};
    },

    getInitialState() {
      return {
          selectedLemmaEdition : "",
          lineLetterValue: "",
      };
    },

    mixins: [ReactMeteorData],

    getMeteorData() {

        var that = this;

        var lemmaText = [];
        // var commentGroup = this.props.commentGroup;
        var selectedLemmaEdition = {
            lines: [],
            slug: ""
        };
        var lemmaQuery = {};
        if (this.props.selectedLineFrom <= this.props.selectedLineTo) {
            lemmaQuery = {
                'work.slug': this.props.workSlug,
                'subwork.n': this.props.subworkN,
                'text.n': {
                    $gte: this.props.selectedLineFrom,
                    $lte: this.props.selectedLineTo
                }
            };
        } else {
            lemmaQuery = {
                'work.slug': this.props.workSlug,
                'subwork.n': this.props.subworkN,
                'text.n': {
                    $gte: this.props.selectedLineFrom,
                    $lte: this.props.selectedLineFrom,
                }
            };
        };

        var textNodesSubscription = Meteor.subscribe('textNodes', lemmaQuery);
        if (textNodesSubscription.ready()) {
            var textNodes = TextNodes.find(lemmaQuery).fetch();
            var editions = [];

            var textIsInEdition = false;
            textNodes.forEach(function(textNode) {

                textNode.text.forEach(function(text) {
                    textIsInEdition = false;

                    editions.forEach(function(edition) {

                        if (text.edition.slug === edition.slug) {
                            edition.lines.push({
                                html: text.html,
                                n: text.n
                            });
                            textIsInEdition = true;
                        }
                    })

                    if (!textIsInEdition) {
                        editions.push({
                            title: text.edition.title,
                            slug: text.edition.slug,
                            lines: [{
                                html: text.html,
                                n: text.n
                            }],
                        })
                    }
                });
            });

            lemmaText = editions;

            if (this.state.selectedLemmaEdition.length) {
                lemmaText.forEach(function(edition) {
                    if (edition.slug === that.state.selectedLemmaEdition) {
                        selectedLemmaEdition = edition;
                    }
                });
            } else {
                selectedLemmaEdition = lemmaText[0];
            }

        }

        return {
            lemmaText: lemmaText,
            selectedLemmaEdition: selectedLemmaEdition
        };

    },

    toggleEdition(editionSlug) {
        if (this.state.selectedLemmaEdition !== editionSlug) {
            this.setState({
                selectedLemmaEdition: editionSlug
            });
        }
    },

    onLineLetterValueChange(event) {
        this.setState({
            lineLetterValue: event.target.value,
        });
    },

    render() {
        var self = this;

        return (
					<div className="comments lemma-panel-visible">
            <div className="comment-outer comment-lemma-comment-outer">

                {this.props.selectedLineFrom > 0 ?
                    <article className="comment lemma-comment paper-shadow">

                        {this.data.selectedLemmaEdition.lines.map(function(line, i){
                            return (
                                <p
                                    key={i}
                                    className="lemma-text"
                                    dangerouslySetInnerHTML={{ __html: line.html}}
                                ></p> );

                        })}

                        {self.props.selectedLineTo === 0 ?
                            <div>
                                <TextField
                                    name="lineLetter"
                                    id="lineLetter"
                                    required={false}
                                    floatingLabelText="Line letter..."
                                    value={this.state.lineLetterValue}
                                    onChange={this.onLineLetterValueChange}
                                />
                            </div>
                            :
                            ""
                        }

                        <div className="edition-tabs tabs">
                            {this.data.lemmaText.map(function(lemmaTextEdition, i){
                                let lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

                                return (<RaisedButton
                                        key={i}
                                        label={lemmaEditionTitle}
                                        data-edition={lemmaTextEdition.title}
                                        className={self.data.selectedLemmaEdition.slug ===  lemmaTextEdition.slug ? "edition-tab tab selected-edition-tab" : "edition-tab tab"}
                                        onClick={self.toggleEdition.bind(null, lemmaTextEdition.slug)}
                                    >
                                    </RaisedButton>);

                            })}
                        </div>

                        <div className="context-tabs tabs">

                            {/* <RaisedButton
                                className="context-tab tab"
                                onClick={this.props.openContextReader}
                                label="Context"
                                labelPosition="before"
                                icon={<FontIcon className="mdi mdi-chevron-right" />}
                            /> */}

                        </div>

                    </article>
                :

                    <article className="comment lemma-comment paper-shadow">
                        <p className="lemma-text no-lines-selected">No line(s) selected</p>
                        {/*<div className="context-tabs tabs">
                            <RaisedButton
                                className="context-tab tab"
                                onClick={this.props.openContextReader}
                                label="Context"
                                labelPosition="before"
                                icon={<FontIcon className="mdi mdi-chevron-right" />}
                                >
                            </RaisedButton>
                        </div>*/}
                    </article>
                }

            </div>
					</div>
        );
    }
});
