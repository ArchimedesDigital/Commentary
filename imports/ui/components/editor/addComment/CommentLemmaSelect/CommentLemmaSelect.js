import React from 'react';
import { Meteor } from 'meteor/meteor';
import { createContainer } from 'meteor/react-meteor-data';
import RaisedButton from 'material-ui/RaisedButton';
import TextField from 'material-ui/TextField';

import getMuiTheme from 'material-ui/styles/getMuiTheme';

// api
import TextNodes from '/imports/api/collections/textNodes';

// lib:
import muiTheme from '/imports/lib/muiTheme';
import Utils from '/imports/lib/utils';


const CommentLemmaSelect = React.createClass({

	propTypes: {
		workSlug: React.PropTypes.string.isRequired,
		subworkN: React.PropTypes.number.isRequired,
		selectedLineFrom: React.PropTypes.number.isRequired,
		selectedLineTo: React.PropTypes.number.isRequired,
		selectedLemmaEdition: React.PropTypes.object,
		lemmaText: React.PropTypes.array,
	},

	childContextTypes: {
		muiTheme: React.PropTypes.object.isRequired,
	},

	getInitialState() {
		return {
			selectedLemmaEdition: '',
			lineLetterValue: '',
		};
	},

	getChildContext() {
		return { muiTheme: getMuiTheme(muiTheme) };
	},

	onLineLetterValueChange(event) {
		this.setState({
			lineLetterValue: event.target.value,
		});
	},

	toggleEdition(editionSlug) {
		if (this.state.selectedLemmaEdition !== editionSlug) {
			this.setState({
				selectedLemmaEdition: editionSlug,
			});
		}
	},

	render() {
		const self = this;

		return (
			<div className="comments lemma-panel-visible">
				<div className="comment-outer comment-lemma-comment-outer">

					{this.props.selectedLineFrom > 0 &&
						this.props.selectedLemmaEdition &&
						'lines' in this.props.selectedLemmaEdition ?
							<article className="comment lemma-comment paper-shadow">

								{this.props.selectedLemmaEdition.lines.map((line, i) => (
									<p
										key={i}
										className="lemma-text"
										dangerouslySetInnerHTML={{ __html: line.html }}
									/>
							))}

								{self.props.selectedLineTo === 0 ?
									<div>
										<TextField
											name="lineLetter"
											id="lineLetter"
											required={false}
											floatingLabelText="Line letter..."
											value={this.state.lineLetterValue}
											onChange={this.onLineLetterValueChange}
										/>
									</div>
								:
								''
							}

								<div className="edition-tabs tabs">
									{this.props.lemmaText.map((lemmaTextEdition, i) => {
										const lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

										return (<RaisedButton
											key={i}
											label={lemmaEditionTitle}
											data-edition={lemmaTextEdition.title}
											className={self.props.selectedLemmaEdition.slug === lemmaTextEdition.slug ?
											'edition-tab tab selected-edition-tab' : 'edition-tab tab'}
											onClick={self.toggleEdition.bind(null, lemmaTextEdition.slug)}
										/>);
									})}
								</div>

								<div className="context-tabs tabs">

									{/* <RaisedButton
								 className="context-tab tab"
								 onClick={this.props.openContextReader}
								 label="Context"
								 labelPosition="before"
								 icon={<FontIcon className="mdi mdi-chevron-right" />}
								 /> */}

								</div>

							</article>
						:

							<article className="comment lemma-comment paper-shadow">
								<p className="lemma-text no-lines-selected">No line(s) selected</p>
								{/* <div className="context-tabs tabs">
							 <RaisedButton
							 className="context-tab tab"
							 onClick={this.props.openContextReader}
							 label="Context"
							 labelPosition="before"
							 icon={<FontIcon className="mdi mdi-chevron-right" />}
							 >
							 </RaisedButton>
							 </div>*/}
							</article>
					}

				</div>
			</div>
		);
	},
});

const CommentLemmaSelectContainer = createContainer(({ selectedLineFrom, selectedLineTo, workSlug, subworkN }) => {
	const lemmaText = [];
	// var commentGroup = this.props.commentGroup;
	const selectedLemmaEdition = {
		lines: [],
		slug: '',
	};
	let lemmaQuery = {};
	if (selectedLineFrom <= selectedLineTo) {
		lemmaQuery = {
			'work.slug': workSlug,
			'subwork.n': subworkN,
			'text.n': {
				$gte: selectedLineFrom,
				$lte: selectedLineTo,
			},
		};
	} else {
		lemmaQuery = {
			'work.slug': workSlug,
			'subwork.n': subworkN,
			'text.n': {
				$gte: selectedLineFrom,
				$lte: selectedLineFrom,
			},
		};
	}

	const textNodesSubscription = Meteor.subscribe('textNodes', lemmaQuery);
	const textNodes = TextNodes.find(lemmaQuery).fetch();
	const editions = [];

	let textIsInEdition = false;
	textNodes.forEach((textNode) => {
		textNode.text.forEach((text) => {
			textIsInEdition = false;

			editions.forEach((edition) => {
				if (text.edition.slug === edition.slug) {
					edition.lines.push({
						html: text.html,
						n: text.n,
					});
					textIsInEdition = true;
				}
			});

			if (!textIsInEdition) {
				editions.push({
					title: text.edition.title,
					slug: text.edition.slug,
					lines: [{
						html: text.html,
						n: text.n,
					}],
				});
			}
		});
	});

	return {
		lemmaText: editions,
		selectedLemmaEdition: editions[0],
		ready: textNodesSubscription.ready(),
	};
}, CommentLemmaSelect);

export default CommentLemmaSelectContainer;
