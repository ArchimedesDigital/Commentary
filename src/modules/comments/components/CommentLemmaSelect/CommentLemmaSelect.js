import React, { Component } from 'react';
import { compose } from 'react-apollo';
import TextField from 'material-ui/TextField';
import PropTypes from 'prop-types';

// graphql
import { editionsQuery } from '../../../../graphql/methods/editions';


class CommentLemmaSelect extends Component {

	constructor(props) {
		super(props);
		this.state = {
			selectedLemmaEdition: '',
			lineLetterValue: '',
		};
		this.onLineLetterValueChange = this.onLineLetterValueChange.bind(this);
		this.toggleEdition = this.toggleEdition.bind(this);

	}

	areQueriesStillLoadingOrTextNodesFromParent(props) {
		let ret = false;
		if(!props.textNodes) {
			ret = true;
		} else if (props.editionsQuery.loading &&
			!(props.lineFrom !== this.props.lineFrom
			|| props.lineTo !== this.props.lineTo)) {
			ret = true;
		}
		return ret;
	}

	componentWillReceiveProps(nextProps) {
		if (this.areQueriesStillLoadingOrTextNodesFromParent(nextProps)) {
			return;
		}

		const editions = nextProps.textNodes;

		this.setState({
			// lemmaText: editions,
			selectedLemmaEdition: editions[0],
		});
	}
	onLineLetterValueChange(event) {
		this.setState({
			lineLetterValue: event.target.value,
		});
	}

	toggleEdition(editionSlug) {
		if (this.state.selectedLemmaEdition !== editionSlug) {
			this.setState({
				selectedLemmaEdition: editionSlug,
			});
		}
	}
	render() {
		const self = this;
		const { selectedLemmaEdition } = this.state;
		return (
			<div className="comments lemma-panel-visible">
				<div className="comment-outer comment-lemma-comment-outer">
					{selectedLemmaEdition && selectedLemmaEdition.lines ?
						<article className="comment lemma-comment paper-shadow">

							{selectedLemmaEdition.lines.map((line, i) => (
								<p
									key={i}
									className="lemma-text"
									dangerouslySetInnerHTML={{ __html: line.html }}
								/>
							))}

							{self.props.lineTo === 0 ?
								<div>
									<TextField
										name="lineLetter"
										id="lineLetter"
										required={false}
										floatingLabelText="Line letter..."
										value={this.state.lineLetterValue}
										onChange={this.onLineLetterValueChange}
									/>
								</div>
								:
								''
							}

							{/* <div className="version-tabs tabs">
									{this.state.lemmaText.map((lemmaTextEdition, i) => {
										const lemmaEditionTitle = Utils.trunc(lemmaTextEdition.title, 20);

										return (<RaisedButton
											key={i}
											label={lemmaEditionTitle}
											data-edition={lemmaTextEdition.title}
											className={self.state.selectedLemmaEdition.slug === lemmaTextEdition.slug ?
											'version-tab tab selected-version-tab' : 'version-tab tab'}
											onClick={self.toggleEdition.bind(null, lemmaTextEdition.slug)}
										/>);
									})}
								</div> */}

							<div className="context-tabs tabs" />
						</article>
						:

						<article className="comment lemma-comment paper-shadow">
							<p className="lemma-text no-lines-selected">No line(s) selected</p>
						</article>
					}
				</div>
			</div>
		);
	}
}

CommentLemmaSelect.propTypes = {
	textNodes: PropTypes.array,
	editionsQuery: PropTypes.object,
};

export default compose(
	editionsQuery
)(CommentLemmaSelect);
