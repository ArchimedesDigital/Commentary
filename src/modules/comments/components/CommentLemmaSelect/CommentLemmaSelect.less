@import '../../../../less/variables.import.less';


.lemmaTextSelected, .lemmaTextUnselected {
  font-family: @font-lemma;
  margin-left: -2px;
}

.lemmaTextUnselected {
  color: #aaa;
}

.lemmaTextSelected {
  color: @text-primary;
}
