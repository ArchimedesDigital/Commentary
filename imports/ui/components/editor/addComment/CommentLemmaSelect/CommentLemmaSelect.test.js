import React from 'react';
import renderer from 'react-test-renderer';

// component:
import CommentLemmaSelect from './CommentLemmaSelect';

describe('CommentLemmaSelect', () => {
	it('renders correctly', () => {

		const tree = renderer
			.create(
				<CommentLemmaSelect
					workSlug={'iliad'}
					subworkN={1}
					selectedLineFrom={1}
				/>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
