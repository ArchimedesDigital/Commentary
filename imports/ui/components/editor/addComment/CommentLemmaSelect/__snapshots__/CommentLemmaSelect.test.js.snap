// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentLemmaSelect renders correctly 1`] = `
<div
  className="comments lemma-panel-visible"
>
  <div
    className="comment-outer comment-lemma-comment-outer"
  >
    <article
      className="comment lemma-comment paper-shadow"
    >
      <p
        className="lemma-text no-lines-selected"
      >
        No line(s) selected
      </p>
    </article>
  </div>
</div>
`;
